### PR TITLE
Infer serializer from collection name for empty results root

### PR DESCRIPTION
**Pull request by mchitten**
#### Purpose

In the event that a collection object is empty and has a name, it's possible to determine the serializer from the name of the object. This may not always be the case -- this could raise `NameError` for a class that does not exist, or `NoSerializerError` if its matching serializer does not exist, in which case this falls back to the original `name.underscore` implementation.
#### Changes

Updates `CollectionSerializer` to try to determine serializer from object name.
#### Related GitHub issues

Possibly(?) a fix for #1745.

### CODE REVIEW

File: lib/active_model/serializer/collection_serializer.rb [hunk 1]
```diff
@@ -44,8 +44,16 @@ def json_key
           explicit_serializer_class._type
         # 2. get from first serializer instance in collection
         key ||= (serializer = serializers.first) && serializer.json_key
-        # 3. get from collection name, if a named collection
-        key ||= object.respond_to?(:name) ? object.name && object.name.underscore : nil
+        if object.respond_to?(:name)
+          begin
+            # 3. use json_key from serializer from collection name
+            key ||= serializer_from_resource(object.name.safe_constantize.new, ActiveModel::Serializer, options).json_key
+          rescue NoSerializerError, NameError
```
Comment (bf4) [inline lib/active_model/serializer/collection_serializer.rb]: Sort of wishing I had merged https://github.com/rails-api/active_model_serializers/pull/1767 at this point...

File: lib/active_model/serializer/collection_serializer.rb [hunk 2]
```diff
@@ -44,8 +44,16 @@ def json_key
           explicit_serializer_class._type
         # 2. get from first serializer instance in collection
         key ||= (serializer = serializers.first) && serializer.json_key
-        # 3. get from collection name, if a named collection
-        key ||= object.respond_to?(:name) ? object.name && object.name.underscore : nil
+        if object.respond_to?(:name)
```
Comment (bf4) [inline lib/active_model/serializer/collection_serializer.rb]:
```ruby
         key ||= (serializer = serializers.first) && serializer.json_key
         # 3. get from collection name, if a named collection
        key ||= begin
          object_name = object.respond_to?(:name) ? object.name : nil
          if (object_class = object_name && object_name.safe_constantize)
            serializer_class = serializer_from_resource(object_class.new, ActiveModel::Serializer, options) { nil }
          end
          serializer_class.nil? : object_name.underscore : serializer_class.json_key
        end
```

just thinking aloud...

```diff
     def serializer_from_resource(resource, serializer_context_class, options)
        serializer_class = options.fetch(:serializer) { serializer_context_class.serializer_for(resource) }

        if serializer_class.nil? # rubocop:disable Style/GuardClause
-          fail NoSerializerError, "No serializer found for resource: #{resource.inspect}"
+          block_given? ? yield : 
+            fail NoSerializerError, "No serializer found for resource: #{resource.inspect}"
        else
          serializer_class.new(resource, options.except(:serializer))
        end
      end
```

Comment (mchitten) [inline lib/active_model/serializer/collection_serializer.rb]: This is great, and definitely a better idea than the `begin...rescue`. I had a couple of tweaks, but updated with more or less the same.

Comment (bf4) [inline lib/active_model/serializer/collection_serializer.rb]: Thanks, was a bit messy as I wrote it... what do you think about changing to throw/catch?

File: CHANGELOG.md [hunk 3]
```diff
@@ -9,6 +9,7 @@ Features:
 Fixes:
 
 - [#1833](https://github.com/rails-api/active_model_serializers/pull/1833) Remove relationship links if they are null (@groyoh)
+- [#1867](https://github.com/rails-api/active_model_serializers/pull/1867) Infer collection root from collection name if possible (@mchitten)
```
Comment (bf4) [inline CHANGELOG.md]: this is really something like 

> Infer collection root preferably from the serializer for the collection name then the collection name, if a named collection

?

File: test/collection_serializer_test.rb [hunk 4]
```diff
@@ -1,5 +1,15 @@
...
 
+class SerializedResource < ActiveModelSerializers::Model
+  # See test line 94
+end
+
+class SerializedResourceSerializer < ActiveModel::Serializer
+  def json_key
+    'resource'
+  end
+end
```
Comment (bf4) [inline test/collection_serializer_test.rb]: Seems you are wanting these in the global scope. Can we perhaps dynamically create and remove these classes or alias exiting ones? (.e.g SerializedResource = Post...)

File: test/collection_serializer_test.rb [hunk 5]
```diff
@@ -23,6 +33,11 @@ def build_named_collection(*resource)
         resource
       end
 
+      def build_named_collection_with_matching_serializer(*resource)
+        resource.define_singleton_method(:name) { 'SerializedResource' }
```
Comment (bf4) [inline test/collection_serializer_test.rb]: do you have a real-world example of this?  Previously there was a test for a monkey-patched array instance and it made me unsure if this was a real feature

File: lib/active_model/serializer/collection_serializer.rb [hunk 6]
```diff
@@ -73,7 +80,8 @@ def serializers_from_resources
       def serializer_from_resource(resource, serializer_context_class, options)
         serializer_class = options.fetch(:serializer) { serializer_context_class.serializer_for(resource) }
 
-        if serializer_class.nil? # rubocop:disable Style/GuardClause
+        if serializer_class.nil?
+          yield if block_given?
```
Comment (bf4) [inline lib/active_model/serializer/collection_serializer.rb]: For this pattern to work, it must either yield or raise.  As written it will yield then raise

```ruby
block_given? and yield or
           fail NoSerializerError, "No serializer found for resource: #{resource.inspect}"
```

would work, I think (not sure about the `and`)

File: lib/active_model/serializer/collection_serializer.rb [hunk 7]
```diff
@@ -45,7 +45,14 @@ def json_key
...
         # 3. get from collection name, if a named collection
-        key ||= object.respond_to?(:name) ? object.name && object.name.underscore : nil
+        key ||= begin
+                  object_name = object.respond_to?(:name) ? object.name : nil
+                  if (object_name && object_class = object_name.safe_constantize)
+                    serializer_class = serializer_from_resource(object_class.new, ActiveModel::Serializer, options) { nil }
+                  end
+                  serializer_class.nil? ? (object_name && object_name.underscore) : serializer_class.json_key
+                end
+
```
Comment (bf4) [inline lib/active_model/serializer/collection_serializer.rb]: ping me when you're done futzing with this. back to work :)

File: test/collection_serializer_test.rb (hunk 5 above)
Comment (mchitten) [inline test/collection_serializer_test.rb]: When I first came across this weirdness, it was actually while testing, and on our API I wasn't seeing the issue. I just did some more poking around and found that it might actually have something to do with the `#none` method.

In this case, `tags` is the expected json_key as defined in the `ResourceTagSerializer`.

```
[1] pry(main)> ActiveModel::Serializer::CollectionSerializer.new(ResourceTag.none, {}).json_key
=> "resource_tags"
[2] pry(main)> ActiveModel::Serializer::CollectionSerializer.new(ResourceTag.last(3), {}).json_key
=> "tags"
```

File: lib/active_model/serializer/collection_serializer.rb (hunk 7 above)
Comment (mchitten) [inline lib/active_model/serializer/collection_serializer.rb]: Sorry -- was fighting an uphill battle with rubocop's `Metrics/PerceivedComplexity`. Based on your comment line 38 I figured it might be okay to disable that for now? Let me know if that's misguided and you'd prefer me to break out into other methods.

File: lib/active_model/serializer/collection_serializer.rb (hunk 2 above)
Comment (mchitten) [inline lib/active_model/serializer/collection_serializer.rb]: I think the `throw` / `catch` idea makes a lot of sense. Does it make sense to throw that in here too? With the exception of the `yield` as you mentioned, I feel like it doesn't actually affect this code anymore sans-`rescue`.

**Comment by mchitten** [comment]
@bf4 Ready for another look when you get the chance

**Comment by mchitten** [comment]
Hey @bf4, any additional thoughts on this? Thanks!

File: test/collection_serializer_test.rb [hunk 8]
```diff
@@ -23,6 +23,21 @@ def build_named_collection(*resource)
...
+        serializer = Class.new(ActiveModel::Serializer) do
+          def json_key
+            'resources'
+          end
+        end
+
+        Object.const_set(:SerializedResource, serialized_resource)
+        Object.const_set(:SerializedResourceSerializer, serializer)
+
+        resource.define_singleton_method(:name) { 'SerializedResource' }
```
Comment (bf4) [inline test/collection_serializer_test.rb]: 1. I'd rather this be in a block so that we can also teardown the classes after defining them
1. I think we can consolidate the code here a little

```ruby
def with_serializable_named_model(name)
  model = Class.new(::Model) do
    def self.name
      'SerializableResource'
    end
  end
  Object.const_set(:SerializableResource, model)
  serializer = Class.new(ActiveModel::Serializer) do
    def json_key
      name
    end
  end
  Object.const_set(:SerializableResourceSerializer, serializer)
  yield model
ensure
  Object.send(:remove_const, :SerializableResource)
  Object.send(:remove_const, :SerializableResourceSerializer)
end
```

then

```ruby
def test_json_key_with_resource_with_name_and_serializer
  with_serializable_named_collection('resources') do |model|
     serializer_instance = collection_serializer.new([model])
     assert_equal 'resources', serializer_instance.json_key
  end
end
```

Comment (bf4) [inline test/collection_serializer_test.rb]: (would also be good to note for the reviewer how the test would fail before the code changes)

Comment (bf4) [inline test/collection_serializer_test.rb]: Not sure how important to your use case, but note that in the test I changed 

```diff
- collection_serializer.new(model)
+ collection_serializer.new([model])
```

since the collection serializer should take in an array of models, or at least something enumerable.  Was that a bug in the test or in your implementation?

**Comment by NullVoxPopuli** [comment]
needs rebase.

@bf4 can you summarize all your comments so @mchitten can finish this? :-)

**Comment by dyoganand** [comment]
Will this be merged??

**Comment by bf4** [comment]
@dyoganand This needed some revisions. Would you be interested in taking it on?

**Comment by bf4** [comment]
same issue: https://github.com/rails-api/active_model_serializers/issues/2087#issuecomment-290205455

**Comment by IzumiSy** [comment]
When is this PR going to be merged? I am being bothered the bug which this PR potentially fixes up exactly at the moment.

**Comment by samuelpismel** [comment]
2 years... Two years! And this isn't merged...

**Comment by NullVoxPopuli** [comment]
@samuelpismel at this point, if you want newish support / a library that easier to maintain, I'd use http://jsonapi-rb.org/
if you don't have a jsonapi style api, jbuilder is probably good enough.

This project has gone through so many rewrites...
The master branch has also been cleared out "in prep" for the next rewrite -- so this PR would need to target the 0.10.x branch.

:-\

Apologies for the circumstances. :(

**Comment by samuelpismel** [comment]
Thanks man, you are correct, sorry about that.
